src/xtt.c

```diff
@@ -247,27 +248,102 @@ enftun_xtt_handshake(const char *server_ip,
     }
 }
 
-static int read_tls_root_cert(TSS2_TCTI_CONTEXT *tcti_context, const char* ca_cert_file)
+static int read_tls_root_cert(unsigned char* tls_root_cert, const char* ca_cert_file, uint16_t tls_len)
 {
-    //read in TLS root cert from TPM
-    unsigned char tls_root_cert[461];
-    int nvram_ret = read_nvram(tls_root_cert,
-                               461,
-                               TLS_ROOT_CERT_HANDLE,
-                               tcti_context);
-    if (0 != nvram_ret) {
-        fprintf(stderr, "Error reading tls root from TPM NVRAM\n");
-        return TPM_ERROR;
-    }
     //write TLS root cert to ca_cert_file
-    int ret = xtt_save_to_file(tls_root_cert, 461, ca_cert_file);
+    int ret = xtt_save_to_file(tls_root_cert, tls_len, ca_cert_file);
     if (ret < 0) {
         return SAVE_TO_FILE_ERROR;
     }
 
     return 0;
 }
 
+static
+int read_in_from_TPM(TSS2_TCTI_CONTEXT* tcti_context,
+                  unsigned char* basename,
+                  uint16_t* basename_len,
+                  xtt_daa_group_pub_key_lrsw* gpk,
+                  xtt_daa_credential_lrsw* cred,
+                  xtt_root_certificate* root_certificate,
+                  unsigned char* tls_root_cert,
+                  uint16_t* tls_len
+                  )
+{
+    size_t sapi_ctx_size = Tss2_Sys_GetContextSize(0);
+    TSS2_SYS_CONTEXT *sapi_context = malloc(sapi_ctx_size);
+    uint16_t length_read = 0;
+    int nvram_ret = 0;
+    nvram_ret = initialize_sapi(sapi_context, sapi_ctx_size, tcti_context);
+    if (0 != nvram_ret) {
+        fprintf(stderr, "Error creating SAPI context\n");
+        goto finish;
+    }
+
+    nvram_ret = xtt_read_object(basename,
```

Comment:
This function is still full of boilerplate.  In my opinion, more of this belongs in a higher-level, object-oriented API in XTT itself.  For example, the final code here could look like:

```
rc = xtt_basename_load(basename, basename_len, sapi_context);
if (rc != 0)
    goto finish;

rc = xtt_daa_group_pub_key_lrsw_load(gpk, sapi_context);
if (rc != 0)
    goto finish;

rc = xtt_daa_credential_lrsw_load(cred, sapi_context);
if (rc != 0)
    goto finish;

rc = xtt_root_certificate_load(root_certificate, sapi_context);
if (rc != 0)
    goto finish;

rc = xtt_tls_root_cert_load(tls_root_cert, tls_len, sapi_context);
if (rc != 0)
    goto finish;
```

Reply:
You mean having, for example, `xtt_basename_load_fromTPM(...)` and `xtt_basename_load_fromFile(...)` methods, that initialize a `basename` "object"?

Reply:
Yes, or maybe `*_load_file` and `*_load_tpm` to keep consistent snakecase and be a little bit shorter.

Reply:
This is related to [this issue in xtt](https://github.com/xaptum/xtt/issues/88), correct?

I agree. But, also agree that we should go ahead with this PR, and discuss that OO design after.